### PR TITLE
Refactor text painting

### PR DESCRIPTION
**Pull request by marijnvdwerf**

### CODE REVIEW

File: src/openrct2/drawing/Text.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    RIGHT
+};
+
+typedef struct TextPaint
+{
+    uint8 colour;
+    sint16 spriteBase;
+    bool underlineText;
+    TextAlignment alignment;
+} Paint;
```
Comment (IntelOrca) [inline src/openrct2/drawing/Text.h]: You do not need to use `typedef` in C++.

File: src/openrct2/drawing/Text.h [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+#pragma once
+
+#include "../common.h"
+
+extern "C" {
+    #include "drawing.h"
+}
+
+enum TextAlignment
```
Comment (IntelOrca) [inline src/openrct2/drawing/Text.h]: `enum class TextAlignment`

File: src/openrct2/drawing/Text.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+{
+    LEFT,
+    CENTER,
```
Comment (IntelOrca) [inline src/openrct2/drawing/Text.h]: `CENTRE`

**Comment by marijnvdwerf** [comment]
There are three string drawing functions remaining:
- `gfx_draw_string_left_centred`: only used for checkbox labels. No real benefit to refactor: doesn't wrap, so ultimately won't be required.
- `draw_string_centred_raw`: draws centred multiline text without wrapping, only 2 usages
- `gfx_draw_string_centred_wrapped_partial`: horizontally and vertically centred, wrapped and only draws a limited amount of characters. Only used for the news ticker, would like to postpone tackling that until we refactor the bottom toolbar.

File: src/openrct2/drawing/Text.cpp [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+{
+
+    utf8 buffer[256];
+
+    format_string(buffer, sizeof(buffer), format, args);
+    drawText(dpi, x, y, paint, buffer);
+}
+
+static void
+drawStringCompat(rct_drawpixelinfo *dpi, sint32 x, sint32 y, rct_string_id format, void *args, uint8 colour,
```
Comment (rwjuk) [inline src/openrct2/drawing/Text.cpp]: Not sure why these methods are called `drawString...`, and the other is `drawText()`. We should use either 'string' or 'text' for consistency.

File: src/openrct2/drawing/Text.cpp [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+    _legacyPaint.underlineText = underline;
+    _legacyPaint.colour = colour;
+    _legacyPaint.alignment = alignment;
+    _legacyPaint.spriteBase = FONT_SPRITE_BASE_MEDIUM;
+    gCurrentFontSpriteBase = FONT_SPRITE_BASE_MEDIUM;
+    drawText(dpi, x, y, &_legacyPaint, format, args);
+}
+
+static void
+drawStringEllipsizedCompat(rct_drawpixelinfo *dpi, sint32 x, sint32 y, sint32 width, rct_string_id format, void *args,
```
Comment (rwjuk) [inline src/openrct2/drawing/Text.cpp]: Ellipsised - en-GB.

File: src/openrct2/drawing/Text.cpp [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+extern "C"
+{
+    #include "../localisation/localisation.h"
+}
+
+static TextPaint _legacyPaint;
+
+void drawText(rct_drawpixelinfo *dpi, float x, float y, TextPaint *paint, utf8string text);
+void drawText(rct_drawpixelinfo *dpi, float x, float y, TextPaint *paint, rct_string_id format, void *args);
```
Comment (IntelOrca) [inline src/openrct2/drawing/Text.cpp]: Can you use TitleCase for these?

File: src/openrct2/drawing/Text.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    CENTRE,
+    RIGHT
+};
+
+struct TextPaint
+{
+    uint8 colour;
+    sint16 spriteBase;
+    bool underlineText;
+    TextAlignment alignment;
```
Comment (IntelOrca) [inline src/openrct2/drawing/Text.h]: Use TitleCase for struct fields.

File: src/openrct2/drawing/Text.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#pragma once
+
+#include "../common.h"
+
+extern "C"
```
Comment (janisozaur) [inline src/openrct2/drawing/Text.h]: remove this `extern`

File: src/openrct2/drawing/Text.h [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    uint8           Colour;
+    sint16          SpriteBase;
+    bool            UnderlineText;
+    TextAlignment   Alignment;
+};
+
+class StaticLayout
+{
+private:
+    utf8string  _buffer;
```
Comment (janisozaur) [inline src/openrct2/drawing/Text.h]: Default constructor should be hidden. I think copy constructor should be hidden too, there are only two instances so far that can easily be converted to non-`auto` variables and it doesn't look like this class is meant to be copied around.

Comment (marijnvdwerf) [reply src/openrct2/drawing/Text.h]: How to do these things?

Comment (janisozaur) [reply src/openrct2/drawing/Text.h]: Just declare them in the private part of class declaration.

```cpp
class Static layout
{
private:
StaticLayout();
StaticLayout (const StaticLayout &);
```

Comment (marijnvdwerf) [reply src/openrct2/drawing/Text.h]: That causes this error to appear:

```
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/drawing/Text.cpp:217:23: error: calling a private constructor of class 'StaticLayout'
        auto layout = StaticLayout(buffer, _legacyPaint, width);
                      ^
/Users/Marijn/Projects/OpenRCT2/Main/src/openrct2/drawing/Text.h:47:5: note: declared private here
    StaticLayout(const StaticLayout &);
```

Comment (janisozaur) [reply src/openrct2/drawing/Text.h]: Yes, that's why I said you have to convert them to regular, non-`auto` variables that don't copy.

Comment (marijnvdwerf) [reply src/openrct2/drawing/Text.h]: Ah, my bad! Completely missed that.

**Pull request merged**